Ruby on Rails Blog introduction

SanGeeky opened the pull request:
I created 2 models, Articles and Comments with their respective crud.

SanGeeky committed:
MVC for Article.

SanGeeky committed:
Create for Article

SanGeeky committed:
Update article

SanGeeky committed:
delete Article

SanGeeky committed:
MVC for Comments

SanGeeky committed:
Concerns, status

SanGeeky committed:
Delete comments and articles.

mauriciosabogal commented:
@SanGeeky Remember to use descriptive messages for commits. For example:

`Add index action on Articles controller`

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+class ArticlesController < ApplicationController
+  
+  http_basic_authenticate_with name: "user", password: "secret", except: [:index, :show]
```
- What is the purpose of this method `http_basic_authenticate_with`?
- Use double quotes only when you need String Interpolation.
- Use `%i` notation for arrays of symbols.

SanGeeky replied:
1. The purpose is making a kind of authentication, with the user called "user" and a password. for some resources in this case the restriction is applied to _Create_, _Edit_ and Delete _Articles_.
2. I will change the strings to a single quote.
3. I review the use of %i and I will change it.

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  http_basic_authenticate_with name: "user", password: "secret", except: [:index, :show]
+
+  def index
+    @articles = Article.all
```
Why don't you just return `Article.all` instead of assigning it to the `@articles` variable?

SanGeeky replied:
I followed the guide, so i will change the return value with `Article.all`.

mauriciosabogal replied:
@SanGeeky what happens if you do that change?

SanGeeky replied:
@mauriciosabogal The values cannot be referenced by the view and @articles  now is a nil variable. So I have to return the @article variables

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def show
+    @article = Article.find(params[:id])
```
What happens if there is no `Article` with the specified `id`? What should you do if that's the case?

SanGeeky replied:
I reviewed this situation and throws an error when id is not found. In this case the solution is redirect to the 404-page error.

SanGeeky replied:
```
class ApplicationController < ActionController::Base
  def not_found
    raise ActionController::RoutingError.new('Not Found')
  end
end
```

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+    @article = Article.find(params[:id])
+  end
+
+  def new
```
What is the difference between `new` and `create` actions? Why do you need them both?

SanGeeky replied:
With `new` , rails return an empty object to be created. And `create` check the necessary values to be stored in the object. I need both because if I only use `create` this will check `params` and raise an error because the form has not been complete yet.

mauriciosabogal replied:
What else? There are more differences.

SanGeeky replied:
Another difference is that `create` with the `.save` method, stores the values in database, and if an error ocurrs render again the form with their respective errors.

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def create
+    # @article = Article.new(title: "Amazing Animals", body: "Show of top 10 Animals"title: "Amazing Animals", body: "Show of top 10 Animals")
```
Remember always to remove commented code.

SanGeeky replied:
Did it !

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  def create
+    # @article = Article.new(title: "Amazing Animals", body: "Show of top 10 Animals"title: "Amazing Animals", body: "Show of top 10 Animals")
+    @article = Article.new(article_params)
+    if @article.save
```
What happens if we use `@article.save!` ?

SanGeeky replied:
Reading documentation about `ActiveRecord` and `save` method, with the bang expressión, executes validations and if a value is missing or wrong raise an exception.

mauriciosabogal replied:
What should you do if it raises an exception?

SanGeeky replied:
In this case, i changed from a conditional to a rescue statement, and works fine!
```
begin
      @article.save!
rescue
      render :new
else
      redirect_to(article_path(@article))
 end
```

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def destroy
+    @article = Article.find(params[:id])
```
You are fetching the article in several actions `Article.find(params[:id])`. What can you do to avoid repetition in that case?

SanGeeky replied:
Create a `Concern` for controllers which return the object if it was found. I will create the concern

mauriciosabogal replied:
Why a `Concern`?

SanGeeky replied:
Because `Article.find(params[:id])` is used in comments_controller many times too, but the other solution is to make a method that returns the found Article. I think creating a Concern for these controllers is a good solution.

SanGeeky replied:
```
module Visible
  extend ActiveSupport::Concern

  def find(id)
    Article.find(id)
  end
end
```

SanGeeky replied:
Final solution
```
before_action :find_article, except: %i[index show]
.....
private
def find_article
    @article = Article.find(params[:id])
end
```

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+
+  def destroy
+    @article = Article.find(params[:id])
+    @article.destroy
```
What happens if you try to delete an `Article` that has several `Comments` associated?

SanGeeky replied:
Because _Article_ and _Comment_ have a `has_many` relationship, the article will be deleted together with their associated comment.

mauriciosabogal replied:
How can you tell it will delete the `Comments` from the database?

SanGeeky replied:
In the model, the parameter `dependent: :destroy` will make a DROP CASCADE query that destroys foreign keys, also the @article object selected and deleted contains the comments instances so will delete all instances.

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  private
+
+  def article_params
+    params.require(:article).permit(:title, :body, :status)
```
- What happens if `params` is an empty `hash`?

- What happens if params is: 

```ruby
{
  article: {
    title: 'An article',
    author: 'John Doe'
  }
}
```

SanGeeky replied:
I tested different cases, when i give an empty hash to  `.permit()` rejects parameters from _Article_:
```
#<ActionController::Parameters {} permitted: true>
Unpermitted parameters: :title, :body, :status
```
If article is an empty hash throws the next error:

`ActionController::ParameterMissing (param is missing or the value is empty: article`

Or if params is an empty hash throws a similar error:

`ActionController::ParameterMissing (param is missing or the value is empty: article):`

Finally with the given params, only accepts the **title** key because the permit method match with :title, :body and :status

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,25 @@
+class CommentsController < ApplicationController
+  
+  http_basic_authenticate_with name: "user", password: "secret", only: :destroy 
```
Remember -> single quotes.

SanGeeky replied:
Done!

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,25 @@
+    redirect_to(article_path(@article))
+  end
+
+  private 
```
Remember to remove trailing spaces. You can configure your editor to do it automatically.

SanGeeky replied:
In this case the code will be written like this ?

   ```
      redirect_to(article_path(@article))
end
private 
```

mauriciosabogal replied:
Nope, trailing spaces are empty spaces at the end of the lines.

The empty line at `line:19` is ok.

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,53 @@
+class ArticlesController < ApplicationController
+  
```
Remember to remove empty lines after `class` definition.

SanGeeky replied:
Done !

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,25 @@
+class CommentsController < ApplicationController
+  
```
Empty line here.

SanGeeky replied:
Done

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,25 @@
+  http_basic_authenticate_with name: "user", password: "secret", only: :destroy 
+  
+  def create
+    @article = Article.find(params[:article_id])
```
Where does the `article_id` param come from?

SanGeeky replied:
Came from _ArticlesController_ get the show method is called and the instance variable @article is defined after in comments form is referenced by this: 

` model: [ @article, @article.comments.build ]`

mauriciosabogal replied:
`ArticlesController` in independent from `CommentController`.

Clue: What is the URL for this action?

SanGeeky replied:
Of course! came from the routes file, and `resources`, so `article_id` is taken from the actual URL 'article/:id' and build by params.

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,25 @@
+    @comment = @article.comments.find(params[:id])
+    @comment.destroy
+
+    redirect_to(article_path(@article))
```
Try not to mix different coding styles:
`redirect_to article_path(@article)` line 9
`redirect_to(article_path(@article))` line 17

SanGeeky replied:
Thank you iI'll keep it in mind

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,25 @@
+
+  def destroy
+    @article = Article.find(params[:article_id])
+    @comment = @article.comments.find(params[:id])
```
How can you find the `Comment` without finding the `Article` in the first place? What is the best approach and why?

SanGeeky replied:
I can delete the comment directly using the class method find:

`@comment = Comment.find(params[:id])`

The best approach is finding the article and deleting the comment because this ensures that we deleted the comment that belongs to the article and finally returns the updated article.

SanGeeky replied:
If i use
`@comments = Comment.find(params[:id])` 
i only make a query to delete the comment, with `params[:id_article]` is verified if this comment belong to the article.

SanGeeky replied:
```
def destroy
    if @comment.article_id == params[:article_id].to_i
      @comment.destroy
    end
    redirect_back fallback_location: articles_url
 end
```

mauriciosabogal commented on app/helpers/articles_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module ArticlesHelper
```
What is the purpose of this empty module?

SanGeeky replied:
This module was created by default with the generate controller command:

`rails generate controller ArticleController Articles`

mauriciosabogal replied:
Remember to be careful when using `generators`, they sometimes create files or code we don't need.

Do you think this `module` is necessary?

SanGeeky replied:
At this moment no, so i must consider the command with the next parameters

` --no-helper`

mauriciosabogal commented on app/helpers/comments_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module CommentsHelper
```
What is the purpose of this empty module?

SanGeeky replied:
This module also was created by default with the generate controller command:

rails generate controller CommandController Commands

mauriciosabogal replied:
Same from here https://github.com/SanGeeky/rails-training/pull/1#discussion_r610291732.

SanGeeky replied:
Answer: [#1 Comment about Helpers](https://github.com/SanGeeky/rails-training/pull/1#discussion_r610642205)

mauriciosabogal commented on app/models/article.rb:
```diff
@@ -0,0 +1,8 @@
+
+  validates :title, presence: true
+  validates :body, presence: true, length: { minimum: 10 }
+end
```
Remember always to put an empty line at the end of the files.

SanGeeky replied:
Got it :smile:

mauriciosabogal commented on app/models/article.rb:
```diff
@@ -0,0 +1,8 @@
+
+  has_many :comments, dependent: :destroy
+
+  validates :title, presence: true
```
When do these validations run?

SanGeeky replied:
When the form of the article new or edit, is sent to the controller when the model attempt to save, check if `:title` has a value if not, returns an array of errors that will be displayed on the form.

mauriciosabogal commented on app/models/comment.rb:
```diff
@@ -0,0 +1,12 @@
+  include Visible
+  belongs_to :article
+
+  VALID_STATUSES = ['public', 'private', 'archived']
```
- Use `%w` notation for arrays of strings.
- Is it ok to define this `VALID_STATUSES` constant inside this model? What happens if you remove it?

SanGeeky replied:
No, because in the Concern file this operation is already done, and if i remove it, model continues works fine.

mauriciosabogal commented on app/models/comment.rb:
```diff
@@ -0,0 +1,12 @@
+
+  validates :status, inclusion: { in: VALID_STATUSES }
+
+  def archived?
```
Is it ok to define this `archived?` method inside this model? What happens if you remove it?

SanGeeky replied:
Same answer from [#1 Comment about Valid Status](https://github.com/SanGeeky/rails-training/pull/1#discussion_r610636657), so i removed

mauriciosabogal commented on app/models/concerns/visible.rb:
```diff
@@ -0,0 +1,18 @@
+module Visible
+  extend ActiveSupport::Concern
```
Why did you create this concern? What is the benefit here?

SanGeeky replied:
The benefit is that with a concern we can drive and share actions on many classes and create mixins to apply the DRY principle. In this case _visible_ will be used by __Article__ and __Comment__

mauriciosabogal commented on app/views/articles/_form.html.erb:
```diff
@@ -0,0 +1,27 @@
+  <div>
+    <%= form.label :title %><br>
+    <%= form.text_field :title %>
+    <% article.errors.full_messages_for(:title).each do |message| %>
```
What happens if `errors` is empty?

SanGeeky replied:
if erros is empty `[]` the .each method iterate zero items and, zero errors will be displayed.

mauriciosabogal commented on app/views/articles/_form.html.erb:
```diff
@@ -0,0 +1,27 @@
+
+  <div>
+    <%= form.label :status %><br>
+    <%= form.select :status, ['public', 'private', 'archived'], selected: 'public' %>
```
The same Ruby rules apply to ERB files. You can use `%w` notation for declaring an array of strings.

SanGeeky replied:
Got it! %w for an array of strings and %i for an array of symbols.

mauriciosabogal commented on app/views/articles/edit.html.erb:
```diff
@@ -0,0 +1,4 @@
+<h1>Edit Article</h1>
+
+<%= render "form", article:@article %>
```
What does this line mean?

SanGeeky replied:
Is the view for Edit Article, so this view will use and render the `_form.html.erb` template and send the variable  `article` which is an internal variable in form template and uses the instance variable from __ArticlesController__.

mauriciosabogal commented on app/views/articles/index.html.erb:
```diff
@@ -0,0 +1,22 @@
+  <% @articles.each do |article| %>
+    <% unless article.archived? %>
+    <li>
+    <%# <a href="/articles/<%=article.id %>
```
Remember to remove commented code.

mauriciosabogal commented on app/views/articles/index.html.erb:
```diff
@@ -0,0 +1,22 @@
+  <% end %>
+</ul>
+
+<%= link_to "New Article", new_article_path %>
```
Remember single quotes here.

mauriciosabogal commented on app/views/articles/new.html.erb:
```diff
@@ -0,0 +1,3 @@
+<h1>New Article</h1>
+
+<%= render "form", article:@article %>
```
Remember single qoutes and the empty line here.

mauriciosabogal commented on app/views/articles/show.html.erb:
```diff
@@ -0,0 +1,18 @@
+<ul>
+  <li><%= link_to "Edit", edit_article_path(@article) %></li>
+  <li>
+    <%= link_to "Destroy", article_path(@article), method: :delete, data: { confirm: "Are you sure?" } %>
```
What is the purpose of this option `method: :delete` here?

SanGeeky replied:
sends as parameter the method which will be used in controller, also for create the url which will be understood by `routes.rb`

mauriciosabogal commented on app/views/articles/show.html.erb:
```diff
@@ -0,0 +1,18 @@
+<% end %>
+
+<h2>Add a comment:</h2>
+<%= render 'comments/form' %>
```
Remember single quotes and the empty line at the end of the file.

SanGeeky replied:
done

mauriciosabogal commented on app/views/comments/_comment.html.erb:
```diff
@@ -0,0 +1,13 @@
+
+<p>
+  <%= link_to 'Destroy Comment', [comment.article, comment], method: :delete, data: { confirm: "Are you sure?" } %>
+</p>
```
Remember single quotes and the empty line at the end of the file.

SanGeeky replied:
Done

mauriciosabogal commented on app/views/comments/_form.html.erb:
```diff
@@ -0,0 +1,19 @@
+
+  <p>
+    <%= form.label :status %><br>
+    <%= form.select :status, ['public', 'private', 'archived'], selected: 'public' %>
```
`%w` notation for arrays.

SanGeeky replied:
Done!

mauriciosabogal commented on config/routes.rb:
```diff
@@ -1,3 +1,10 @@
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+  root 'articles#index'
+  # Route _____ controller#action
+  # get '/articles', to: 'articles#index'
```
Remember to remove commented code.

SanGeeky replied:
Got it!

mauriciosabogal commented on config/routes.rb:
```diff
@@ -1,3 +1,10 @@
+  # get '/articles', to: 'articles#index'
+  # get 'articles/:id', to: "articles#show"
+  resources :articles do
+    resources :comments
```
- Why are the `comments` resources inside `articles`?
- Where are all the routes for `articles` and `comments`?
- Do you think using `resources :comments` is the best approach here? Why?

SanGeeky replied:
1. Because comments belongs to an article, if comments are individuals this could go outside from `articles`
2. The routes are created by the resources method, and that block with comments adds the respective url for comments.
3. Yes if i need to update a comment or show the selected comment. But in this case, i can add only the necessary routes, like this:
```
resources :articles do
     resources :comments, only: %i[create destroy]
end
```

mauriciosabogal reviewed (changes requested):
Remember to apply code styling guides to all the files.
You can install the `rubocop` gem and configure the linter of your editor.

mauriciosabogal commented on app/models/comment.rb:
```diff
@@ -0,0 +1,12 @@
+
+  VALID_STATUSES = ['public', 'private', 'archived']
+
+  validates :status, inclusion: { in: VALID_STATUSES }
```
Is it ok to define this validation here?

SanGeeky replied:
Same answer from [#1 Comment about Valid Status](https://github.com/SanGeeky/rails-training/pull/1#discussion_r610636657), so i removed

SanGeeky committed:
Reformating code on articles controllers

- Filters applied to find_article.
- `save!` method used to verify the record on db.
- Not found exception when article is not found.

SanGeeky committed:
Reformating code on comments controller.

- Filters used to find comments and articles individualy
-  Find article to create comments
- Finding a comment to destroy the respective comment

SanGeeky committed:
Reformating views with Rubocop code style.

SanGeeky committed:
Reformating Models with rubocop code style.

SanGeeky committed:
Resources for commets limited to, *create* and *destroy*

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,60 @@
+  end
+
+  def show
+    @article 
```
Is it necessary to return the `@article` variable?

SanGeeky replied:
I tested the method without return the `@article` variable and works fine, but I have a question. Should I delete the `show`method ?

mauriciosabogal replied:
Yes, you can delete it. That's because you're using `resources :articles` in the routes file. You can read about this [here](https://guides.rubyonrails.org/layouts_and_rendering.html#rendering-by-default-convention-over-configuration-in-action).

SanGeeky replied:
Thank you for the explanation!

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,60 @@
+  end
+
+  def edit
+    @article
```
Is it necessary to return the @article variable?

SanGeeky replied:
[Same answer](https://github.com/SanGeeky/rails-training/pull/1#discussion_r611911896)

Should I delete the `edit` method? Because at this moment the method only returns the article.

mauriciosabogal replied:
See https://github.com/SanGeeky/rails-training/pull/1#discussion_r612080487

SanGeeky replied:
Thanks!

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,60 @@
+    @article = Article.new(article_params)
+    begin
+      @article.save!
+    rescue StandardError => e
```
You can delete the error object if you are not going to use it.

You can use `ActiveRecord::RecordInvalid` instead of `StandardError`.

SanGeeky replied:
Is correct to put the error message in the console? 
```
rescue ActiveRecord::RecordInvalid => error
      puts error.message
      render :new 
```

mauriciosabogal replied:
The correct way to do it is to use the Rails Logger in case there is the need to log the error.

SanGeeky replied:
Got it, useful to debug  [Ruby Debug](https://guides.rubyonrails.org/debugging_rails_applications.html#the-logger)

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,60 @@
+    rescue StandardError => e
+      render :new
+    else
+      redirect_to(article_path(@article))
```
Why did you include the `else` here?

SanGeeky replied:
I consider this from [How to Rescue Exceptions](https://stackify.com/rescue-exceptions-ruby/)
![image](https://user-images.githubusercontent.com/24459154/114455298-353a8680-9ba1-11eb-904d-a3b5e3a72fe7.png)

Should I use the `redirect_to` by default without use `else` ?

```
def create
  ....
  redirect_to article_path(@article)
end
```

mauriciosabogal replied:
See https://github.com/SanGeeky/rails-training/pull/1/files#r612082199

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,60 @@
+  def find_article
+    begin
+      @article = Article.find(params[:id])
+    rescue => exception
```
You can delete the error object if you are not going to use it. Also, always specify an error class if possible.

SanGeeky replied:
In this case i use this:
```
def find_article
    @article = Article.find(params[:id])
  rescue ActiveRecord::RecordNotFound => e
    puts e.message

```

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,60 @@
+  def create
+    @article = Article.new(article_params)
+    begin
+      @article.save!
```
Why are you using the bang method for saving the article (`create` action), and not for updating or deleting the article?

SanGeeky replied:
I read this [Misuse update/update! in Ruby on Rails](https://marcgg.com/blog/2020/04/08/rails-update-bang/), and to maintain a standard I will use the bang method and handle the exceptions when I need to record to the database.

mauriciosabogal replied:
@SanGeeky Here, in this case, you can use both `save` or `save!` (the same for `update` or `update!`, etc.). The option you chose depends on how you are going to handle the errors. If you are going to do only a redirection when an error occurs while saving the object, you can use `save`. 

Using the bang methods like `save!` makes sense if you need to perform more actions when the error occurs (like logging the error messages, rollbacks, event notifications, or trigger other processes), or if you have an error handler for all the controllers.

SanGeeky replied:
Thanks :+1: in this case, now i consider only using methods without the bang method. If a Record fails will return false and redirect to the respective view. Example in `ArticlesController`

```ruby
def create
    @article = Article.new(article_params)
    @article.save ? redirect_to(@article) : render(:new)
end
```

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  def find_article
+    @article = Article.find(params[:article_id])
```
Is it necessary to fetch the `article` from the database?

SanGeeky replied:
No, is not necessary. I made this operation:

```
def create
    @comment = Comment.new(comments_params)
    @comment.article_id = params[:article_id]
    begin
      @comment.save!
    rescue ActiveRecord::RecordInvalid => e
      puts e.message
    end
    redirect_back fallback_location: articles_url
 end
```

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,32 @@
+  http_basic_authenticate_with name: 'user', password: 'secret', only: :destroy
+
+  def create
+    @comment = @article.comments.create(comments_params)
```
Is it necessary to create the `comment` using the `article`?

SanGeeky replied:
[create comment](https://github.com/SanGeeky/rails-training/pull/1#discussion_r611967401)

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -0,0 +1,32 @@
+  end
+
+  def destroy
+    if @comment.article_id == params[:article_id].to_i
```
Please refactor this to make the code cleaner.

SanGeeky replied:
Done !

mauriciosabogal commented:
@SanGeeky There are some files containing trailing whitespace.

SanGeeky committed:
Update methods in Controllers, and handle errors.

- Show and Edit only use find_article method to return article.
-  Update and Destroy now are using the bang method.
- comment now is created with Comment class.

Removed trailing spaces from files.

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -1,32 +1,31 @@
-
-  def find_article
-    @article = Article.find(params[:article_id])
+    params.require(:comment).permit(:commenter, :body, :status, :article_id)
```
`article_id` is not part of the comment body. You can remove it.

SanGeeky replied:
Done !

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -1,32 +1,31 @@
+      @comment.save!
+    rescue ActiveRecord::RecordInvalid
+    end
+    redirect_back fallback_location: articles_url
```
This should be inside the `begin` block, like 

```ruby
    begin
      @comment.save!
       redirect_back fallback_location: articles_url
    rescue ActiveRecord::RecordInvalid
    end

```

SanGeeky replied:
Update of `create` method on `CommentsController`
```ruby
def create
    @comment = Comment.new(comment_params)
    @comment.save
    redirect_article
end
```

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -1,32 +1,31 @@
+    @comment.article_id = params[:article_id]
+    begin
+      @comment.save!
+    rescue ActiveRecord::RecordInvalid
```
Rescuing from an error makes sense if you're handling the errors. Don't leave this empty.

SanGeeky replied:
okay !

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -1,32 +1,31 @@
-    @comment = @article.comments.create(comments_params)
-    redirect_to article_path(@article)
+    @comment = Comment.new(comments_params)
+    @comment.article_id = params[:article_id]
```
You can build the params hash, including the `article_id`, in the `comments_params` method. That way, you don't need to assign the `article_id` to  `@comment`.

SanGeeky replied:
I made this operation and returning a hash, is correct doing this ?
```ruby
def comments_params
    comment = params.require(:comment).permit(:commenter, :body, :status).to_h
    comment['article_id'] = params['article_id']
    comment
end
```

SanGeeky replied:
Update from `comment_params`
```ruby
def comment_params
    comment = params.require(:comment).permit(:commenter, :body, :status)
    comment.merge(article_id: params[:article_id])
end
```

mauriciosabogal commented on app/controllers/comments_controller.rb:
```diff
@@ -1,32 +1,31 @@
-    if @comment.article_id == params[:article_id].to_i
-      @comment.destroy
-    end
+    @comment.destroy if @comment.article_id.eql? params[:article_id].to_i
```
You can find the `@comment` using its `id` and the `article_id` in the same query. That way you don't need to do this comparison here.

SanGeeky replied:
I made this query:
`@comment = Comment.find_by(id: params[:id], article_id: params[:article_id])`

SanGeeky replied:
if returns `nil` i made a ternary conditional:
`@comment.nil? ? not_found : @comment`

mauriciosabogal replied:
Here, this method assigns the found comment to the variable `@comment`, so it is not necessary to return it. Consider using the`if`modifier like this:

`not_found if @comment.nil?`

SanGeeky replied:
Got it! also, i will keep  in mind these little things

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,57 @@
+    begin
+      @article.save!
+    rescue ActiveRecord::RecordInvalid
+      puts e.message
```
this will raise an error because `e` does not exist.

SanGeeky replied:
My bad, now it's deleted

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,57 @@
+      puts e.message
+      render :new
+    end
+    redirect_to @article
```
This should be inside the `begin` block, like this:

```ruby
    begin
      @article.save!
      redirect_to @article
    rescue ActiveRecord::RecordInvalid
```

SanGeeky replied:
Done ! and thanks for the recommendation

SanGeeky replied:
Update of `save` method: [Update Save Method](https://github.com/SanGeeky/rails-training/pull/1#discussion_r612730149)

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,57 @@
+    rescue ActiveRecord::RecordInvalid
+      render :edit
+    end
+    redirect_to @article
```
Same as https://github.com/SanGeeky/rails-training/pull/1/files#r612082199

SanGeeky replied:
Update `update` method
```ruby
def update
    @article.update(article_params) ? redirect_to(@article) : render(:edit)
end
```

mauriciosabogal commented on app/controllers/articles_controller.rb:
```diff
@@ -0,0 +1,57 @@
+    rescue ActiveRecord::RecordNotDestroyed
+      redirect_back fallback_location: root_path
+    end
+    redirect_to root_path
```
Same as https://github.com/SanGeeky/rails-training/pull/1/files#r612082199

SanGeeky replied:
Update Destroy Method
```ruby
def destroy
    @article.destroy
    redirect_to(root_path)
end
```

SanGeeky committed:
Controllers updated rescues and query for comments

SanGeeky committed:
comments_params renamed to comment_param

SanGeeky committed:
Articles and Comments Controllers methods updated.
-  save, update, destroy without bang method in Articles
- save destroy methods at this momment always return to redirect_article
-  use of conditional instead of rescues.
-  when a comment is no found return nil and raise not_found

SanGeeky commented:
With the recommendations, now controllers are easier to read. For the next PR I will keep in mind add validations for comments.

SanGeeky committed:
find_comment updated not_found condition.
